src/ParticleDA.jl

```diff
@@ -430,89 +180,127 @@ function run_particle_filter(init, filter_params::FilterParameters, model_params
+            )
         end
 
         # Broadcast resampled particle indices to all ranks
         MPI.Bcast!(filter_data.resampling_indices, filter_params.master_rank, MPI.COMM_WORLD)
-
-        @timeit_debug timer "get_particles" particles = get_particles(model_data)
-        @timeit_debug timer "State Copy" copy_states!(particles,
-                                                      filter_data.copy_buffer,
-                                                      filter_data.resampling_indices,
-                                                      my_rank,
-                                                      nprt_per_rank)
-
-        @timeit_debug timer "Mean and Var" get_mean_and_var!(filter_data.statistics, particles, filter_params.master_rank)
+    
+        @timeit_debug timer "Copy states" copy_states!(
+            states,
+            filter_data.copy_buffer,
+            filter_data.resampling_indices,
+            my_rank,
+            nprt_per_rank
+        )
+                                                      
+        if filter_params.verbose
+            @timeit_debug timer "Update statistics" update_statistics!(
+                filter_data.statistics, states, filter_params.master_rank
+            )
+        end
 
         if my_rank == filter_params.master_rank && filter_params.verbose
 
-            @timeit_debug timer "IO" begin
-                unpack_statistics!(filter_data.avg_arr, filter_data.var_arr, filter_data.statistics)
-                write_snapshot(filter_params.output_filename, model_data, filter_data.avg_arr, filter_data.var_arr, filter_data.weights, it)
-            end
+            @timeit_debug timer "Unpack statistics" unpack_statistics!(
+                filter_data.unpacked_statistics, filter_data.statistics
+            )
+            @timeit_debug timer "Write snapshot" write_snapshot(
+                filter_params.output_filename,
+                model,
+                filter_data,
+                states,
+                time_index,
+                time_index in filter_params.particle_save_time_indices,
+            )
 
         end
 
     end
+    
+    if !filter_params.verbose
+        # Do final update and unpack of statistics if not performed in filtering loop
+        @timeit_debug timer "Update statistics" update_statistics!(
+            filter_data.statistics, states, filter_params.master_rank
+        )
+        if my_rank == filter_params.master_rank
+            @timeit_debug timer "Unpack statistics" unpack_statistics!(
+                filter_data.unpacked_statistics, filter_data.statistics
+            )
+        end
+    end
+    
+    # Gather final states from all MPI ranks to master
+    @timeit_debug timer "Gather states" flat_global_states = MPI.Gather(
+        vec(states), filter_params.master_rank, MPI.COMM_WORLD
+    )
+    if my_rank == filter_params.master_rank
+        states = reshape(flat_global_states, (size(states, 1), filter_params.nprt))
+    end
```

Comment:
This currently gathers the state particles from all ranks to the master rank, which in retrospect is probably a bad idea to have as the default behaviour given the high memory overhead when the number of ranks is high as Tuomas pointed out today (given this is done only once the communication overhead should be relatively minor). I made this change partly as having access to the full final set of states is useful from an analysis perspective and to allow potentially continuing filtering, and also so that the summary statistics correspond to the statistics of these returned particles for testing, but I guess it's probably better to allow the user to manually gather the states on the master rank if needed. Posting this as a reminder for me to fix 😁.